pkg/controller/livenessprobemapnodeprobe/probemapnodeprobe_utils.go

```diff
@@ -0,0 +1,223 @@
+package livenessprobemapnodeprobe
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	livenessprobeUtils "github.com/openkruise/kruise/pkg/util/livenessprobe"
+)
+
+func (r *ReconcileEnhancedLivenessProbeMapNodeProbe) delNodeProbeConfig(pod *v1.Pod) error {
+	return r.retryOnConflictDelNodeProbeConfig(pod)
+}
+
+func (r *ReconcileEnhancedLivenessProbeMapNodeProbe) addNodeProbeConfig(pod *v1.Pod) error {
+	plivenessProbeConfig, err := parseEnhancedLivenessProbeConfig(pod)
+	if err != nil {
+		return err
+	}
+	if pod.Spec.NodeName == "" {
+		klog.Warningf("No found pod node name, pod: %v/%v", pod.Namespace, pod.Name)
+		return nil
+	}
+
+	npp := appsv1alpha1.NodePodProbe{}
+	err = r.Get(context.TODO(), types.NamespacedName{Name: pod.Spec.NodeName}, &npp)
+	if err == nil {
+		return r.retryOnConflictAddNodeProbeConfig(pod, plivenessProbeConfig)
+	}
+	if !errors.IsNotFound(err) {
+		klog.Errorf("Failed to get node pod probe for pod: %s/%s, err: %v", pod.Namespace, pod.Name, err)
+		return err
+	}
+	newNppGenerated, err := r.generateNewNodePodProbe(pod)
+	if err != nil {
+		klog.Errorf("Failed to generate new node pod probe for pod: %s/%s, err: %v", pod.Namespace, pod.Name, err)
+		return err
+	}
+	if err = r.Create(context.TODO(), &newNppGenerated); err != nil {
+		klog.Errorf("Failed to create node pod probe for pod: %s/%s, err: %v", pod.Namespace, pod.Name, err)
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcileEnhancedLivenessProbeMapNodeProbe) retryOnConflictDelNodeProbeConfig(pod *v1.Pod) error {
+	nppClone := &appsv1alpha1.NodePodProbe{}
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: pod.Spec.NodeName}, nppClone); err != nil {
+			if errors.IsNotFound(err) {
+				klog.Warningf("No found getting updated npp %s from client", pod.Spec.NodeName)
+				return nil
+			}
+			klog.Errorf("Error getting updated npp %s from client", pod.Spec.NodeName)
+			return err
+		}
+
+		podNewNppCloneSpec := nppClone.Spec.DeepCopy()
+
+		newPodProbeTmp := []appsv1alpha1.PodProbe{}
+		for index := range podNewNppCloneSpec.PodProbes {
+			podContainerProbe := podNewNppCloneSpec.PodProbes[index]
+			if podContainerProbe.Name == pod.Name && podContainerProbe.Namespace == pod.Namespace &&
+				podContainerProbe.UID == fmt.Sprintf("%v", pod.UID) {
+				continue
+			}
+			newPodProbeTmp = append(newPodProbeTmp, podContainerProbe)
+		}
+		podNewNppCloneSpec.PodProbes = newPodProbeTmp
+
+		// delete node pod probe
+		if len(podNewNppCloneSpec.PodProbes) == 0 {
```

Comment:
if node still exists, it seems unnecessary to delete the nodepodprobe, just leave it empty seems enough

Reply:
done, remove 
`
	// delete node pod probe
	if len(podNewNppCloneSpec.PodProbes) == 0 {
		if err := r.Delete(context.TODO(), nppClone); err != nil {
			klog.Errorf("Failed to delete node pod probe for pod: %s/%s, err: %v", pod.Namespace, pod.Name, err)
			return err
		}
		return nil
	}
`